src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,271 @@
+      opt_iframe =
+          moveLayoutRect(opt_iframe, -opt_iframe.left, -opt_iframe.top);
+    }
+
+    // Normalize container LayoutRect to be relative to page
+    let elementRect;
+    let ownerRect = null;
+
+    // If calls with container, always assume getLayoutBox() return relative
+    // LayoutRect to container
+    elementRect = this.element_.getLayoutBox();
+    const owner = this.element_.getOwner();
+    ownerRect = owner && owner.getLayoutBox();
+
+    const changeEntry = this.getValidIntersectionChangeEntry_(
+        elementRect, ownerRect, hostViewport, opt_iframe);
+    if (changeEntry) {
+      this.callback_(changeEntry);
+    }
+  }
+
+  /**
+   * Return a change entry for that should be compatible with
+   * IntersectionObserverEntry if it's valid with current config.
+   * When the new intersection ratio doesn't cross one of a threshold value,
+   * the function will return null.
+   *
+   * Mutates passed in rootBounds to have x and y according to spec.
+   *
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {?./layout-rect.LayoutRectDef} owner element's owner rect
+   * @param {!./layout-rect.LayoutRectDef} hostViewport hostViewport's rect
+   * @param {./layout-rect.LayoutRectDef=} opt_iframe. iframe container rect
+   * @return {?IntersectionObserverEntry} A valid change entry or null if ratio
+   * does not fill in a new threshold bucket.
+   * @private
+   */
+  getValidIntersectionChangeEntry_(element, owner, hostViewport, opt_iframe) {
+    // If opt_iframe is provided, all LayoutRect has position relative to
+    // the iframe.
+    // If opt_iframe is not provided, all LayoutRect has position relative to
+    // the host document.
+
+    // Building an IntersectionObserverEntry.
+    let intersectionRect = element;
+    const nonIntersectRect = layoutRectLtwh(0, 0, 0, 0);
+    // element intersects with its owner.
+    if (owner) {
+      intersectionRect = rectIntersection(owner, element) || nonIntersectRect;
+    }
+    // element intersects with container iframe.
+    if (opt_iframe) {
+      intersectionRect = rectIntersection(opt_iframe, intersectionRect) ||
+          nonIntersectRect;
+    }
+    // element intersects with hostViewport
+    intersectionRect = rectIntersection(hostViewport, intersectionRect) ||
+        nonIntersectRect;
+
+    // calculate ratio, call callback based on new ratio value.
+    const ratio = intersectionRatio(intersectionRect, element);
+    const newThresholdSlot = getThresholdSlot(this.threshold_, ratio);
+    if (newThresholdSlot == this.prevThresholdSlot_) {
+      return null;
+    }
+    this.prevThresholdSlot_ = newThresholdSlot;
+
+    let boundingClientRect = element;
+    let rootBounds = hostViewport;
+
+    // If element inside an iframe. Every Layoutrect has already adjust their
+    // origin according to opt_iframe rect origin.
+    if (opt_iframe) {
+      // If element inside an iframe. Every LayoutRect has already adjust their
```

Comment:
This description is not correct. It's because the iframe can't scroll, so the viewport is at (0, 0), so the element is already relative to the viewport (since they're relative to (0, 0)).